azkaban-common/src/main/java/azkaban/executor/ExecutionControllerUtils.java

```diff
@@ -100,11 +120,68 @@ public static void finalizeFlow(final ExecutorLoader executorLoader, final Alert
       logger.error("Failed to finalize flow " + flow.getExecutionId() + ", do not alert user.", e);
     }
 
+    // If the flow is in state EXECUTION_STOPPED and user enabled restartability in flow parameters
+    if (flow.getStatus() == Status.EXECUTION_STOPPED) {
+      final ExecutionOptions executionOptions = flow.getExecutionOptions();
+      if (executionOptions != null) {
+        final Map<String, String> flowParam = executionOptions.getFlowParameters();
+        if (flowParam != null && !flowParam.isEmpty()) {
+          if (Boolean.valueOf(flowParam.getOrDefault(FlowParameters
+              .FLOW_PARAM_ALLOW_RESTART_ON_EXECUTION_STOPPED, "false"))) {
+            restartExecutableFlow(flow);
+          }
+        }
+      }
+    }
+
     if (alertUser) {
       alertUserOnFlowFinished(flow, alerterHolder, getFinalizeFlowReasons(reason, originalError));
     }
   }
 
+  /**
+   * When a flow is in EXECUTION_STOPPED state, and user allows to restart the entire flow execution
+   * with same flow options, a new execution will be dispatched
+   * @param exFlow
+   */
+  public static void restartExecutableFlow(final ExecutableFlow exFlow) {
+    if (projectManager == null || executorManagerAdapter == null) {
+      logger.error("ExecutionControllerUtils not properly initialized.");
+      return;
+    }
+    if (exFlow.getDispatchMethod() == DispatchMethod.CONTAINERIZED) { // Enable restartability
+      // for containerized execution
+      final Project project;
+      final Flow flow;
+      try {
+        project = FlowUtils.getProject(projectManager, exFlow.getProjectId());
```

Comment:
Shouldn't we reuse the logic from Prepare execution?

Reply:
Prepare execution restarts from the failed nodes. In this case, we want to do a full restart.